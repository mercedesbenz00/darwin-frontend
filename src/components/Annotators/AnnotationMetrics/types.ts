import { MembershipPayload, DatasetAnnotationReportPayload } from '@/store/types'

export type AnnotationDataRange = 'total' | 'month' | 'week' | 'day'
export type AnnotationDataGranularity = 'day' | 'hour'

export type ChartDimension =
  'annotationTime' |
  'imagesAnnotated' |
  'avgTimePerAnnotation' |
  'annotationsCreated' |
  'reviewPassRate'

export type AdaptedDataPoint =
  {[k in Exclude<ChartDimension, 'reviewPassRate' | 'avgTimePerAnnotation'>]: number } &
  {[k in Extract<ChartDimension, 'reviewPassRate' | 'avgTimePerAnnotation'>]: number | null} &
  { userId: number | null, date: string }

export type AggregatedDataPoint = Pick<
    DatasetAnnotationReportPayload,
    | 'annotation_time'
    | 'annotations_approved'
    | 'annotations_created'
    | 'date'
    | 'images_annotated'
    | 'images_approved'
    | 'images_rejected'
    | 'user_id'
  >

export type AnnotationDimensionData = {
  annotationTime: number,
  annotationsCreated: number,
  avgTimePerAnnotation: number | null,
  imagesAnnotated: number,
  reviewPassRate: number | null
}

export type AnnotationChartData = {
  data: AdaptedDataPoint[]
  dimensions: AnnotationDimensionData
  member?: MembershipPayload
  visible: boolean
}

export type AnnotationBarChartData = {
  data: AggregatedDataPoint[]
  member?: MembershipPayload
}

/**
 * Format for a single item to be shown in a chart tooltip.
 *
 * A tooltip will contain multiple items
 */
export type ChartTooltipData = { member: MembershipPayload | null, value: string}
