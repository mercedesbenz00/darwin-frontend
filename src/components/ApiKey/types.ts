export type AbilityGroup =
  'manage_comments' |
  'manage_tasks' |
  'manage_datasets' |
  'manage_team' |
  'manage_team_members' |
  'manage_models'

export type Group = { id: AbilityGroup; name: string }
export type GroupOption = Group & { selected: boolean }

export enum SupportedLanguages {
  shell = 'shell',
  python = 'python',
  javascript = 'javascript',
  elixir = 'elixir',
  cli = 'cli'
}

export type AccessCodeExample = { language: SupportedLanguages, code: string }
