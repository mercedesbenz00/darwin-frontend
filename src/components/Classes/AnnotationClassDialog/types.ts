import { AnnotationClassPayload, AnnotationClassMetadata, AnnotationTypeName } from '@/store/types'

type PendingFields =
  | 'crop_key'
  | 'crop_url'
  | 'image_height'
  | 'image_width'
  | 'key'
  | 'original_image_url'
  | 'scale'
  | 'x'
  | 'y'

/**
 * Holds a class image that isn't necessarily persisted to the backend yet.
 * That means it might be missing any of the url fields
 *
 * It can be saved to backend once at least the `key` field is set, as that one
 * can be used to infer/compute all the other fields.
 *
 * - `crop_key` - repleace "original" with "crop" in `key` field
 * - `crop_url` - signed s3 url (by backend), for `crop_key`
 * - `original_image_url` - signed s3 url (by backend) for `key`
 */
export type PendingClassImage =
  Omit<AnnotationClassPayload['images'][0], PendingFields>
  & Partial<Pick<AnnotationClassPayload['images'][0], PendingFields>>

// eslint-disable-next-line camelcase
export type PendingClassImageWithUrlSet = PendingClassImage & { original_image_url: string }

export const isClassImage = (
  image: PendingClassImage
): image is AnnotationClassPayload['images'][0] =>
  !!image.key && !!image.crop_key

export type AnnotationClassData = {
  datasets: AnnotationClassPayload['datasets']
  description: string | null
  hotkey: null | string
  id: number | null
  images: PendingClassImage[]
  metadata: AnnotationClassMetadata
  name: string | null
}

export type DefaultAnnotationClassData = Partial<AnnotationClassData> & {
  selectedMainAnnotationType?: AnnotationTypeName
}

export type AnnotationClassValidationErrors = {
  annotationTypes?: string
  name?: string
  skeleton?: string
}

export const VIEWPORT_WIDTH: number = 150
export const VIEWPORT_HEIGHT: number = 150

export type CropInfo = {
  orientation: number
  points: [string, string, string, string]
  zoom: number
}

/**
 * VueCroppie is really archaic and requires us to get crop info via $ref,
 * so this type defines that ref
 */
export type Croppie = {
  /** Binds options to the underlying croppie library */
  bind: (options: Pick<CropInfo, 'points' | 'zoom'> & { url: string}) => void

  /** Retrieves current crop info */
  get: () => CropInfo
  result: (options?: Record<string, string>, callback?: Function) => void

  /** Sets scale/zoom level of cropper */
  setZoom: (value: number) => void
}
