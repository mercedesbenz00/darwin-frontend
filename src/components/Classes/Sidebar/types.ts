import { AnnotationTypeName } from '@/store/types'

export type AnnotationTypeFilterItemType = {
  id: string | number
  label: string
  icon?: string
  name: AnnotationTypeName
  color: string
  count: number | null
  [key: string]: any
}
