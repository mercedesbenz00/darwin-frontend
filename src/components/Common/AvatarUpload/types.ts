export type AvatarUploadData = {
  hash: string
  file: File
  type: string | null
}
