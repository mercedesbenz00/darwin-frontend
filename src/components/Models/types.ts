import {
  RunningSessionPayload,
  TrainedModelPayload,
  TrainingSessionPayload
} from '@/utils/wind/types'

export type ModelItem = {
  datasetSlug: string | null
  id: string
  insertedAt: string
  name: string
  teamId: number | null
  runningSession?: RunningSessionPayload
  trainingSession?: TrainingSessionPayload
  trainedModel?: TrainedModelPayload
}
