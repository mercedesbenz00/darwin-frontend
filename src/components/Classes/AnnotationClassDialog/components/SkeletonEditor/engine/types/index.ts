export * from './SkeletonEdgeType'
export * from './SkeletonNodeType'
