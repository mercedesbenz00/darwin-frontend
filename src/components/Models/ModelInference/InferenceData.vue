<template>
  <div class="inference">
    <div
      ref="canvasContainer"
      class="inference__view"
    >
      <layout-component
        v-if="editor.value"
        :editor="editor.value"
      />
    </div>
    <div class="inference__sidebar">
      <div class="inference__sidebar__section">
        <h4 class="inference__sidebar__section__header">
          Labels
        </h4>
        <div class="inference__labels">
          <div
            v-for="[label, { count, color }] in groupedAnnotations"
            :key="label"
            class="inference__label"
          >
            <div
              class="inference__label__text"
              :style="{background: color}"
            >
              {{ label }}
            </div>
            <div class="inference__label__count">
              {{ count }}
            </div>
          </div>
        </div>
      </div>
      <div class="inference__sidebar__section">
        <div class="inference__sidebar__section__header">
          <h4>Model output</h4>
          <copy-to-clipboard
            class="inference__sidebar__section__header__copy-button"
            :value="json"
          />
        </div>
        <pre
          class="inference__sidebar__section__content inference__json"
        >{{ JSON.stringify(inferenceResults, null, 2) }}</pre>
      </div>
      <div class="inference__sidebar__section">
        <positive-button
          v-if="mode === 'image'"
          class="inference__sidebar__button"
          @click="$emit('repick')"
        >
          Use another image
        </positive-button>
        <negative-button
          v-if="mode === 'video'"
          class="inference__sidebar__button"
          @click="$emit('stop-stream')"
        >
          Stop stream
        </negative-button>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop, Watch, Inject } from 'vue-property-decorator'

import CopyToClipboard from '@/components/Common/CopyToClipboard.vue'
import LayoutComponent from '@/components/WorkView/Layout'
import { Editor } from '@/engine/editor'
import { InferenceResult } from '@/engineCommon/backend'
import { TrainedModelPayload } from '@/store/types'
import { getRGBAColorHash, rgbaString } from '@/utils'

@Component({
  name: 'inference-data',
  components: {
    LayoutComponent,
    CopyToClipboard
  }
})
export default class InferenceData extends Vue {
  @Inject()
  editor!: { value: Editor | null }

  @Prop({ required: false, type: Array, default: () => [] })
  inferenceResults!: InferenceResult[]

  @Prop({ required: false, type: Array, default: () => [] })
  classes!: TrainedModelPayload['classes']

  @Prop({ required: true, type: String })
  imageData!: string

  @Prop({ required: true, type: String })
  mode!: 'image' | 'video'

  canvasSize: { width: number, height: number } | null = null

  $refs!: {
    canvasContainer: HTMLDivElement
    canvas: HTMLCanvasElement
  }

  mounted (): void {
    if (!this.editor.value) { return }
    this.editor.value.init(true)
    this.editor.value.activeView.showPills = false
    this.editor.value.disableEditSubAnnotations()
    const plugins = this.editor.value.pluginManager.pluginsForView()
    this.editor.value.installAllPlugins(plugins)
    // circumvent editor tool selection,
    // since this component doesn't need workview state management
    this.editor.value.toolManager.activateTool('select_tool')
    this.resizeCanvas()
    this.updateEditor()
    this.editor.value.activateCallbacks()

    window.addEventListener('resize', this.resizeCanvas)
    this.$once('hook:beforeDestroy', () => window.removeEventListener('resize', this.resizeCanvas))
  }

  resizeCanvas (): void {
    const { canvasContainer } = this.$refs
    this.editor.value?.layout.updateViewsCameraDimensions(
      canvasContainer.offsetWidth,
      canvasContainer.offsetHeight
    )
    this.editor.value?.scaleToFit()
  }

  @Watch('inferenceResults')
  onInferenceResults (): void { this.updateEditor() }

  updateEditor (): void {
    const { classes, editor, inferenceResults } = this
    if (!editor.value || !inferenceResults) { return }
    editor.value.activeView.overlayManager.reset()
    editor.value.activeView.measureManager.reset()
    editor.value.activeView.setInferenceData(inferenceResults, classes)
  }

  @Watch('imageData', { immediate: true })
  onImageData (imageData: string | null): void {
    if (!imageData) { return }
    this.$nextTick(() => {
      this.editor.value?.activeView.setRawImage(imageData)
    })
  }

  // TODO: Figure out how to get labels (classes)
  get groupedAnnotations () {
    const { classes } = this
    const initial: { [s: string]: { color: string, count: number} } = {}

    const grouping = this.inferenceResults.reduce((acc, annotation) => {
      const inferenceLabel = annotation.label || annotation.name
      if (!inferenceLabel) { return acc }
      const inferenceClass = classes.find(c => c.name === inferenceLabel)

      const displayName = inferenceClass
        ? inferenceClass.display_name || inferenceClass.name
        : inferenceLabel
      if (!displayName) { return acc }
      if (acc[displayName]) {
        acc[displayName].count += 1
      } else {
        const color = rgbaString(getRGBAColorHash(displayName))
        acc[displayName] = { count: 1, color }
      }
      return acc
    }, initial)

    return Object.entries(grouping)
  }

  get json (): string {
    return JSON.stringify(this.inferenceResults, null, 2)
  }
}
</script>

<style lang="scss" scoped>
.inference {
  @include row;
  width: 100%;

  .inference__view {
    flex: 5;
    overflow: hidden;
    position: relative;

    :deep(canvas) {
      @include noSelect;
      width: 100%;
      height: 100%;
    }
  }

  .inference__sidebar {
    flex: 2;
    margin-left: 25px;
    display: grid;
    height: 100%;
    grid-template-rows: 4fr 3fr auto;
    grid-row-gap: 25px;
  }

  .inference__sidebar__section {
    display: grid;
    grid-template-rows: auto 1fr;
    grid-auto-rows: 1fr;
    grid-row-gap: 15px;
    overflow: hidden;

    > :last-child {
      margin-bottom: -15px;
    }
  }

  .inference__labels {
    .inference__label {
      @include row--distributed;
      align-items: baseline;

      &:not(:last-child) {
        margin-bottom: 10px;
      }

      &__text,
      &__count {
        border-radius: 3px;
        @include typography(md-1, default);
        vertical-align: middle;
      }

      &__text {
        padding: 8px;
        color: $colorWhite;
      }

    }
  }

  .inference__sidebar__section__header {
    @include row--distributed;
    @include typography(xl, headlines, bold);
  }

  .inference__json {
    @include typography(md-2, source);
    line-height: 150%;
    background: $colorSecondaryLight3;
    color: $colorSecondaryLight;
    padding: 8px 0 8px 10px;
    white-space: pre-wrap;
    overflow: auto;
  }

  .inference__sidebar__button {
    width: 100%;
  }
}
</style>
