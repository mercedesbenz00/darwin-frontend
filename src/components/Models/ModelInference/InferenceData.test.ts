import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'

import createTestStore from 'test/unit/createTestStore'
import { buildTrainingClass } from 'test/unit/factories'

import { Editor } from '@/engine/editor'
import { ItemManager } from '@/engine/managers'
import { InferenceResult } from '@/engineCommon/backend'
import { installCommonComponents } from '@/plugins/components'
import { TrainedModelPayload } from '@/store/types'

import InferenceData from './InferenceData.vue'

const localVue = createLocalVue()
localVue.use(Vuex)
installCommonComponents(localVue)

let store: ReturnType<typeof createTestStore>
let propsData: {
  classes?: TrainedModelPayload['classes']
  imageData: string
  inferenceResults?: InferenceResult[]
  mode: 'image' | 'video'
}
let editor: Editor

beforeEach(() => {
  store = createTestStore()
  propsData = {
    classes: [],
    imageData: 'fooBase64',
    inferenceResults: [],
    mode: 'image'
  }

  editor = new Editor(new ItemManager(store), store)
  editor.init(true)
})

it('matches snapshot', () => {
  const wrapper = shallowMount(InferenceData, {
    localVue,
    propsData,
    store,
    provide: { editor: { value: editor } }
  })
  expect(wrapper).toMatchSnapshot()
})

it('matches snapshot with annotations', () => {
  propsData.classes = [
    buildTrainingClass({ id: '1', name: 'Foo', type: 'polygon' }),
    buildTrainingClass({ id: '2', name: 'Bar', type: 'bounding_box' })
  ]

  propsData.inferenceResults = [
    { label: 'Foo', path: [] },
    { label: 'Foo', path: [] },
    { label: 'Bar', path: [] }
  ]

  const wrapper = shallowMount(InferenceData, {
    localVue,
    propsData,
    store,
    provide: { editor: { value: editor } }
  })
  expect(wrapper).toMatchSnapshot()
})
