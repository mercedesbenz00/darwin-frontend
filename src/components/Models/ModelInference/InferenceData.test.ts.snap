// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div class="inference">
  <div class="inference__view">
    <layout-component-stub editor="[object Object]"></layout-component-stub>
  </div>
  <div class="inference__sidebar">
    <div class="inference__sidebar__section">
      <h4 class="inference__sidebar__section__header">
        Labels
      </h4>
      <div class="inference__labels"></div>
    </div>
    <div class="inference__sidebar__section">
      <div class="inference__sidebar__section__header">
        <h4>Model output</h4>
        <copy-to-clipboard-stub title="Copy to clipboard" value="[]" class="inference__sidebar__section__header__copy-button"></copy-to-clipboard-stub>
      </div> <pre class="inference__sidebar__section__content inference__json">[]</pre>
    </div>
    <div class="inference__sidebar__section">
      <positive-button-stub class="inference__sidebar__button">
        Use another image
      </positive-button-stub>
      <!---->
    </div>
  </div>
</div>
`;

exports[`matches snapshot with annotations 1`] = `
<div class="inference">
  <div class="inference__view">
    <layout-component-stub editor="[object Object]"></layout-component-stub>
  </div>
  <div class="inference__sidebar">
    <div class="inference__sidebar__section">
      <h4 class="inference__sidebar__section__header">
        Labels
      </h4>
      <div class="inference__labels">
        <div class="inference__label">
          <div class="inference__label__text" style="background: rgb(147, 210, 45);">
            Foo
          </div>
          <div class="inference__label__count">
            2
          </div>
        </div>
        <div class="inference__label">
          <div class="inference__label__text" style="background: rgb(108, 224, 112);">
            Bar
          </div>
          <div class="inference__label__count">
            1
          </div>
        </div>
      </div>
    </div>
    <div class="inference__sidebar__section">
      <div class="inference__sidebar__section__header">
        <h4>Model output</h4>
        <copy-to-clipboard-stub title="Copy to clipboard" value="[
  {
    &quot;label&quot;: &quot;Foo&quot;,
    &quot;path&quot;: []
  },
  {
    &quot;label&quot;: &quot;Foo&quot;,
    &quot;path&quot;: []
  },
  {
    &quot;label&quot;: &quot;Bar&quot;,
    &quot;path&quot;: []
  }
]" class="inference__sidebar__section__header__copy-button"></copy-to-clipboard-stub>
      </div> <pre class="inference__sidebar__section__content inference__json">[
  {
    "label": "Foo",
    "path": []
  },
  {
    "label": "Foo",
    "path": []
  },
  {
    "label": "Bar",
    "path": []
  }
]</pre>
    </div>
    <div class="inference__sidebar__section">
      <positive-button-stub class="inference__sidebar__button">
        Use another image
      </positive-button-stub>
      <!---->
    </div>
  </div>
</div>
`;
